client/device/cloudftd/cloudftdonboarding/create.go

```diff
@@ -0,0 +1,136 @@
+	if err != nil {
+		return nil, err
+	}
+	if len(readFmcDomainRes.Items) == 0 {
+		return nil, fmt.Errorf("failed to read fmc domain uid, fmc domain info not found")
+	}
+	fmcDomainUid := readFmcDomainRes.Items[0].Uuid
+
+	// 2. get a system token for creating FTD device record in FMC
+	// CDO token does not work, we will get a 405 method not allowed if we do that
+	client.Logger.Println("getting a system token for creating FTD device record in FMC")
+
+	// 2.1 get tenant context => tenant uid
+	readOutp, err := user.GetTokenInfo(ctx, client, user.NewGetTokenInfoInput())
+	if err != nil {
+		return nil, err
+	}
+	tenantUid := readOutp.UserAuthentication.Details.TenantUid
+	// 2.2 get a system token with scope of tenant uid
+	createTokenOutp, err := user.CreateSystemToken(ctx, client, user.NewCreateSystemTokenInput(tenantUid))
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. post device record
+	client.Logger.Println("creating FTD device record in FMC")
+
+	// 3.1 read ftd metadata
+	readFtdOutp, err := cloudftd.ReadByUid(ctx, client, cloudftd.NewReadByUidInput(createInp.FtdUid))
+	if err != nil {
+		return nil, err
+	}
+	// 3.2 create ftd device
+	createDeviceInp := fmcconfig.NewCreateDeviceRecordInputBuilder().
+		Type("Device").
+		NatId(readFtdOutp.Metadata.NatID).
+		Name(readFtdOutp.Name).
+		AccessPolicyUid(readFtdOutp.Metadata.AccessPolicyUid).
+		LicenseCaps(readFtdOutp.Metadata.LicenseCaps).
+		PerformanceTier(readFtdOutp.Metadata.PerformanceTier).
+		RegKey(readFtdOutp.Metadata.RegKey).
+		FmcDomainUid(fmcDomainUid).
+		FmcHostname(fmcRes.Host).
+		SystemApiToken(createTokenOutp.AccessToken).
+		Build()
+	var createOutp fmcconfig.CreateDeviceRecordOutput
+	err = retry.Do(
+		fmcconfig.UntilCreateDeviceRecordSuccess(ctx, client, createDeviceInp, &createOutp),
+		retry.NewOptionsBuilder().
+			Retries(-1).
+			Delay(3*time.Second).
+			Timeout(1*time.Hour). // it can take 15-20 minutes for FTD to come up + 10 minutes to create device record
+			Logger(client.Logger).
+			EarlyExitOnError(false).
+			Build(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// 4. trigger FTD onboarding state machine
+	client.Logger.Println("re-triggering FTD onboarding state machine")
+
+	// 4.1 get ftd specific device
+	ftdSpecificOutp, err := cloudftd.ReadSpecific(ctx, client, cloudftd.NewReadSpecificInputBuilder().Uid(readFtdOutp.Uid).Build())
+	if err != nil {
+		return nil, err
+	}
+	// 4.2 trigger register state machine
+	_, err = cloudftd.UpdateSpecific(ctx, client,
+		cloudftd.NewUpdateSpecificFtdInput(
+			ftdSpecificOutp.SpecificUid,
+			"INITIATE_FTDC_REGISTER",
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// 4.3 wait until state machine done
+	err = retry.Do(
+		cloudftd.UntilStateDone(ctx, client, cloudftd.NewReadByUidInput(createInp.FtdUid)),
+		retry.NewOptionsBuilder().
+			Retries(-1).
+			Delay(1*time.Second).
+			Timeout(10*time.Minute).
```

Comment:
I'd give it 15-20 minutes.

Also, what happens if the FTD being onboarded is not a FTDv where we are passing the configure manager command in user data, but a FTD where the user needs to SSH in and manually paste the user data?

Reply:
Sure changed to 20 minutes.
If you are worrying about 20 minutes is not enough for user to enter the command: this onboarding state machine is triggered after we polled and registered the device with FMC, so in the case where user SSH in, they will not be having this 10 minutes limit, instead they will be in the `retry` loop in the above code, which has a timeout of 1 hour.